### PR TITLE
add link checker to validate script

### PR DESCRIPTION
**Pull request by mitulshah-suse**
Related to https://github.com/rancher/rancher/issues/41096

### CODE REVIEW

**Comment by mitulshah-suse** [comment]
The CI is currently failing due to the `20.10.sh` linking pointing to the incorrect version `20.10.21.sh`
This should start passing once the PR https://github.com/rancher/install-docker/pull/111 is merged.

**Comment by mitulshah-suse** [comment]
@kinarashah CI passed now that the PR is merged. Can you please check and approve this.

File: scripts/validate [hunk 1]
```diff
@@ -35,3 +35,17 @@ done
 
 echo "==> Checking basic shell validation"
 find ../dist -name \*.sh -exec sh -n {} \;
+
+
+echo "==> Add link checker to verify all links in dist are pointing to latest versions"
+
+for LINK_PREFIX in `find ../dist/ -type l -exec basename {} \; | grep sh | sed 's/\.sh//' | sort -V`
```
Comment (snasovich) [inline scripts/validate:42]: I'm not too versed in bash scripts, but doesn't `find ../dist/ -type l -exec basename {} \;` already return name without extension?
(Though logs from the run seem to disagree with this, could you please explain?)

File: scripts/validate [hunk 2]
```diff
@@ -35,3 +35,17 @@ done
...
+do
+    if [[ `find ../dist/ -type f -name "$LINK_PREFIX*" -exec basename {} \; | sort -V | tail -1` == `readlink ../dist/$LINK_PREFIX.sh` ]]
+    then
+        echo "$LINK_PREFIX.sh is pointing to the latest `readlink ../dist/$LINK_PREFIX.sh`"
```
Comment (snasovich) [inline scripts/validate:46]: These scripts generally only output errors, but I think it's a good idea to output these as a "sanity check".
Could we prefix it with `[INF]` though?

File: scripts/validate [hunk 3]
```diff
@@ -35,3 +35,17 @@ done
 
 echo "==> Checking basic shell validation"
 find ../dist -name \*.sh -exec sh -n {} \;
+
+
+echo "==> Add link checker to verify all links in dist are pointing to latest versions"
+
+for LINK_PREFIX in `find ../dist/ -type l -exec basename {} \; | grep sh | sed 's/\.sh//' | sort -V`
+do
+    if [[ `find ../dist/ -type f -name "$LINK_PREFIX*" -exec basename {} \; | sort -V | tail -1` == `readlink ../dist/$LINK_PREFIX.sh` ]]
```
Comment (snasovich) [inline scripts/validate:44]: Suggesting introducing variables for these parts as we then reference them in messages produced.

File: scripts/validate [hunk 4]
```diff
@@ -35,3 +35,17 @@ done
...
+    then
+        echo "$LINK_PREFIX.sh is pointing to the latest `readlink ../dist/$LINK_PREFIX.sh`"
+    else
+        echo "[ERROR] $LINK_PREFIX.sh is pointing to `readlink ../dist/$LINK_PREFIX.sh` instead of the latest version `find ../dist/ -type f -name "$LINK_PREFIX*" -exec basename {} \; | sort -V | tail -1`"
```
Comment (snasovich) [inline scripts/validate:48]: Suggesting to use `[ERR]` prefix for consistency with existing errors.

**Review by snasovich — CHANGES_REQUESTED** [review]
Added a few comments to address / explain.

Meanwhile, wondering if there is a way to see output of running this validation script? The one I see in the run is very minimal and doesn't have `echo` messages - https://drone-pr.rancher.io/rancher/install-docker/177/1/2

```
v0.5.4: Pulling from rancher/dapper
Digest: sha256:bd027dda55dacac2a275ec41f816362ee1971ff54747aa9e18697fdd1fd0f5d3
Status: Downloaded newer image for rancher/dapper:v0.5.4
+ find ./dist -name \*.sh -exec sh -n {} \;
```

Actually, looks like validate is not actually running at this step - https://github.com/rancher/install-docker/blob/b24ff15452a9d511dae151dc19110b7f5bdb9135/.drone.yml#L13

It does run as part of `test` step - https://drone-pr.rancher.io/rancher/install-docker/177/1/3 (starting at line 121).

Could you please fix it while at it so `validate` properly runs `validate` step and `test` doesn't run `validate`?

File: scripts/validate (hunk 1 above)
Comment (mitulshah-suse) [reply scripts/validate:42]: `basename -s` would remove the specified suffix, but these commands are not running a bash shell. They are running in a busybox container and all the commands have very limited functionality. I tried with `basename -s .sh` and it kept throwing errors. Newer busybox containers seem to have fixed it, but the version in use here does not have that fix. So this is more like a work in all scenarios sort of solution.

File: scripts/validate (hunk 4 above)
Comment (mitulshah-suse) [reply scripts/validate]: Done

File: scripts/validate (hunk 3 above)
Comment (mitulshah-suse) [reply scripts/validate]: Done

File: scripts/validate (hunk 2 above)
Comment (mitulshah-suse) [reply scripts/validate]: Done

File: scripts/validate (hunk 1 above)
Comment (snasovich) [reply scripts/validate]: @mitulshah-suse , thank you - I should have googled a bit more. :)

**Review by snasovich — APPROVED** [review]
Investigate build failures on Rocky 8, otherwise LGTM

**Comment by mitulshah-suse** [comment]
> Investigate build failures on Rocky 8, otherwise LGTM

@snasovich That was some temp issue. Seems to be fixed on rerun. Can you please merge this PR.

**Comment by mitulshah-suse** [comment]
@kinarashah Can you please review as well.

**Review by superseb — APPROVED** [review]

**Pull request merged**